### PR TITLE
fix: Release metrics return incorrect featuresPerWeek value #280

### PR DESCRIPTION
**Pull request by YuliaMittova**
Bugfix for issue #280

### CODE REVIEW

**Comment by innokenty** [comment]
So if there no bug – then we can't create any dataset from this issue, we just cancel it.

No need to create a PR next time – you can just drop a comment on the issue 👍

**Comment by innokenty** [comment]
So actually.... this isn't a problem with the test, **the bug is valid**! 

Your test mocks the value "now" to comply with the logic, but that is exactly the problem: logic of calculating this percentage of **features per week should not depend on the current time for the releases that are already released**! This is exactly the whole point: when a release already released, this value shouldn't change, no matter how much time has passed since then.

Look at the golden test referred on the bug issue: it actually emulates the correct behaviour for a releases that has released_at date, and not mocking any instants.

**Review by innokenty — CHANGES_REQUESTED** [review]
The bug-issue is valid: please re-analyze, add a bugfix and provide correct tests that validate the logic correctly. Look into the linked golden test – it seems pretty good to me.

**Comment by YuliaMittova** [comment]
> The bug-issue is valid: please re-analyze, add a bugfix and provide correct tests that validate the logic correctly. Look into the linked golden test – it seems pretty good to me.

@innokenty fixed logic and tests. Please take a look.

File: src/test/java/com/sivalabs/ft/features/api/controllers/MetricsIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Test
+    void shouldReturnZeroVelocityWhenLessThanTwoWeeksOfData() throws Exception {
+        assertThat(d(velocity(getMetrics("GO-2024.2.3")), "featuresPerWeek")).isEqualTo(0.0);
+    }
+
+    @Test
+    void shouldReturnZeroVelocityWhenThereIsNoData() throws Exception {
+        assertThat(d(velocity(getMetrics("GO-2024.2.7")), "featuresPerWeek")).isEqualTo(0.0);
+    }
```
Comment (innokenty) [inline src/test/java/com/sivalabs/ft/features/api/controllers/MetricsIntegrationTests.java:54]: Please add more test coverage. The only real test is only checking the most simple case of "5.0" and does not cover all the use cases.

Image
Image
Image

**Review by innokenty — CHANGES_REQUESTED** [review]
We need a bit more tests, than a single basic happy-case scenario, please cover some corner cases.

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java [hunk 2]
```diff
@@ -318,14 +323,14 @@ private Instant calculateEstimatedEndDate(Release release, List<Feature> feature
         return estimatedEndDate.atStartOfDay(ZoneOffset.UTC).toInstant();
     }
 
-    private long calculateBusinessWeeksElapsed(Instant startDate) {
-        long businessDaysElapsed = calculateBusinessDaysElapsed(startDate);
+    private long calculateBusinessWeeksElapsed(Instant startDate, Instant endDate) {
```
Comment (ElizavetaP) [inline src/main/java/com/sivalabs/ft/features/domain/DashboardService.java]: The precision is lost here: long businessDaysElapsed / 5.
Please preserve the fractional business-week value instead of truncating it, and add a test where the period between createdAt and releasedAt includes a partial business week.

**Review by ElizavetaP — CHANGES_REQUESTED** [review]

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java (hunk 2 above)
Comment (innokenty) [reply src/main/java/com/sivalabs/ft/features/domain/DashboardService.java]: yeah, we definitely need a test for a fractional value

File: src/test/java/com/sivalabs/ft/features/api/controllers/MetricsIntegrationTests.java (hunk 1 above)
Comment (YuliaMittova) [reply src/test/java/com/sivalabs/ft/features/api/controllers/MetricsIntegrationTests.java:54]: @innokenty
1. Fixed test shouldReturnZeroVelocityWhenLessThanTwoWeeksOfData
2. Added more tests
3. Added double to the logic to make calculations more precise

File: src/main/java/com/sivalabs/ft/features/domain/DashboardService.java (hunk 2 above)
Comment (YuliaMittova) [reply src/main/java/com/sivalabs/ft/features/domain/DashboardService.java]: Added fraction there

**Review by innokenty — APPROVED** [review]
I added some fixes:

- fixed `shouldReturnZeroVelocityWhenLessThanTwoWeeksOfData` test to actually test
the <2 weeks threshold by – instead of returning 0 due to zero features.
- changed release descriptions in the data table to human-readable comments explaining what each release tests
- added test for fractional featuresPerWeek calculation

**Review by ElizavetaP — APPROVED** [review]
